sohel2020/senior-infra-engineer-task
# Infrastructure Engineer Challenge: Initial Implementation

### Description
This PR implements the full solution for the Infrastructure Engineer Challenge. The key components include:

### Cluster Setup

- Local Kubernetes cluster creation using Kind and Docker Desktop.

### Monitoring and Logging

- Setup of monitoring stack using kube-prometheus-stack via Helmfile (Prometheus, Grafana, Alertmanager, etc.).

- Implementation of log aggregation with Grafana Loki (Promtail, Loki).

### Secret Management

- Encrypted secrets using SOPS and GPG, with .sops.yaml rules for secure storage.

### Application Deployment

- Deployment of a sample guestbook application with Skaffold and Helm.

- Option for custom deployment via interactive Bash scripts.

### Automation

- Integrated pre-commit hooks to enforce consistent coding standards.

### Documentation

- Comprehensive README with step-by-step instructions for setting up and verifying the infrastructure.

### Testing Instructions
1. Follow the README to:

- Install prerequisites.
- Set up the cluster.
- Deploy monitoring and logging stacks.
- Deploy the application and verify its availability.

2. Verify the following:

- Monitoring stack is accessible via Grafana, Prometheus, and Alertmanager UIs.
- Logs are visible in Grafana (Loki Data Source).
- Application is running, and metrics are being collected in Prometheus.

### Checklist

- [x]  Secrets securely managed.
- [x]  Monitoring stack successfully deployed.
- [x]  Logging stack operational.
- [x]  Application deployment verified.
